### DISCUSSION TITLE
-threshold causing images to get outlined

### DISCUSSION THREAD

**Discussion by cellobass** [ImageMagick/ImageMagick#7365] [Help]
### ImageMagick version

7.1.1-Q16-HDRI

### Operating system, version and so on

Windows 10

### Description

I can't get the threshold command to work properly. When I feed in a grayscale image and use -threshold 50%, the result is always that the image appears to be outlined.  I was hoping for black to stay black and white to stay white.  Any idea what the issue could be?  Here is the command line I'm using:

`magick -density 400 input.pdf -threshold 50% -verbose output.pdf`

Original Image:
![image](https://github.com/ImageMagick/ImageMagick/assets/87158556/61204223-8776-4ac5-9688-84b69f04f3e9)

Output Image:
![image](https://github.com/ImageMagick/ImageMagick/assets/87158556/b769c727-db8c-40e9-894f-17c1e691884b)

I suspect it's because they may be 16-bit grayscale.  But if I'm right, how would that change the command line parameters?

Thank you!

[input.pdf](https://github.com/user-attachments/files/15527328/input.pdf)

**Comment by snibgo** [comment]
Please post your actual `input.pdf`, not a PNG from a screenshot of a program that displays your PDF.

**Reply by cellobass** [reply]
Thanks, I've attached it.

**Comment by fmw42** [comment]
Try this:

`magick -density 400 input.pdf -alpha off -auto-threshold otsu x.png`


![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/74e52506-89d0-40e4-897d-841be8845f12)

**Reply by cellobass** [reply]
That's helpful.  I had done the -auto-threshold otsu switch before and obtained similar results.  Your -alpha off switch makes the black reappear.  Upon further analysis in Photoshop, the "outline" image was actually replacing the black pixels with transparent pixels, so the -alpha off switch made them black pixels again.  Now I have a better idea of what was (is) happening.

However, both commands seem to create a grayscale image (with just two levels in it: 0 and 255), not a two-color bitmap file.  I'm trying to reduce the size of the output file by only outputting a black and white image, not something with 8- or 16-bits per channel.  I thought the -threshold switch claimed to do that, but now that I look at the image in Photoshop, it's also doing the same thing.

Is there a way to convert the 8- or 16-bit grayscale image to pure black and white (1-bit image)?  Thank you!

**Reply by fmw42** [reply]
-threshold or -auto-threshold only creates a binary result, which is only 1-bit (grayscale).  Binary means two levels, 0 and 255.  I do not understand what you want?  Do you want two levels with 0 and 1 rather than 0 and 255?  If so, you will not see any difference, it will look totally black.

**Reply by cellobass** [reply]
I've discovered that the move back to grayscale comes when it's converted back to PDF from PNG.  So, ImageMagick is actually outputting a 1-bit image (correctly).  However, if I specify the PDF file format as output instead of PNG, something (GhostScript presumably) is making it grayscale again.  I'm guessing this is a "just the way it works" issue, but one can only hold out hope!

The PNG version is 126KB, the PDF version is 1,279KB (twice the input PDF file size).  Disk space is effectively free, so it doesn't really matter from a disk space standpoint.  However, I've found that large PDF files print much faster if they contain 1-bit images rather than 8- or 16-bit grayscale.  And when I'm printing reams of music, it makes a huge difference in time.  So imagine a 100 page PDF file.  Thanks!

**Reply by fmw42** [reply]
Before writing to PDF, add -type bilevel to your command line.  Does that help?

**Reply by cellobass** [reply]
Interesting!  It cut the PDF file size in half again (to match the original PDF file size).  But it's still a grayscale image in PDF.  Who coulda' guessed?

**Reply by fmw42** [reply]
That is not what I get:

`magick -density 400 input.pdf -alpha off -auto-threshold otsu -type bilevel x.pdf`

```
magick identify x.pdf
Image:
  Filename: x.pdf
  Permissions: rw-r--r--
  Format: PDF (Portable Document Format)
  Mime type: application/pdf
  Class: DirectClass
  Geometry: 629x890+0+0
  Resolution: 72x72
  Print size: 8.73611x12.3611
  Units: Undefined
  Colorspace: sRGB
  Type: Bilevel.     <--------
  Base type: Undefined
  Endianness: Undefined
  Depth: 16/1-bit.   <--------
  Channel depth:   < -------
    red: 1-bit
    green: 1-bit
    blue: 1-bit
    alpha: 1-bit

```

**Reply by cellobass** [reply]
Thank you.  That means that how I'm opening in Photoshop (via Acrobat) is alerting it.  It seems like I have what I need.  

I really appreciate it!
